File: CodingInterviews/LinkList/ReversePrint.java
package LinkList;

import java.util.Stack;

class Solution {
    public int[] reversePrint(ListNode head) {
        if (head == null)
            return new int[0];
        Stack<Integer> stack = new Stack<>();
        while (head != null) {
            stack.push(head.val);
            head = head.next;
        }
        int n = stack.size();
        int[] res = new int[n];
        for (int i = 0; i < n; i++)
            res[i] = stack.pop();
        return res;
    }
}

File: LeetCode/CodingInterviews/reversePrint.go
package coding

func reversePrint(head *ListNode) []int {
	if head == nil {
		return make([]int, 0)
	}
	var res = reversePrint(head.Next)
	res = append(res, head.Val)
	return res
}

func reversePrintRecurrence(head *ListNode) []int {
	var res = make([]int, 0)
	var p = head
	for p != nil {
		res = append(res, p.Val)
		p = p.Next
	}
	// reverse
	var i, j = 0, len(res) - 1
	for i < j {
		res[i], res[j] = res[j], res[i]
		i++
		j--
	}
	return res
}
